Governance Model Proposal (#213)
Qiskit/openqasm: governance.md

The TSC may have a **Secretary**  who organizes the meeting (e.g. prepares and distributes the agenda and collects a set of issues to be considered). The TSC **Secretary** will be a **Contributor** appointed by vote of the TSC.

The TSC will, from time to time, elect to create **Working Groups** to study issues. The TSC will appoint a **Contributor** as chair for each WG. These WGs should come back to the TSC with a proposal in the form of a language RFC or pull request to the OpenQASM specification. A WG will be automatically disbanded upon acceptance or (final) rejection of an RFC.

> **TSC Members** will serve a 2-year term, with the exception of the founding TSC which will have staggered terms to avoid an election of an entire TSC all at once. New **TSC Members** will be selected by nomination of the TSC, followed by a vote which is open to all **Contributors**. Outgoing **TSC Members** participate in the nomination process which selects their successors. Note that **TSC Members** may be re-elected to serve additional 2-year terms. The same process will be used in the even that a **TSC Member** resigns prior to the end of their term.

Reviewer: ```suggestion
**TSC Members** will serve a 2-year term, with the exception of the founding TSC which will have staggered terms to avoid an election of an entire TSC all at once. New **TSC Members** will be selected by nomination of the TSC, followed by a vote which is open to all **Contributors**. Outgoing **TSC Members** participate in the nomination process which selects their successors. Note that **TSC Members** may be re-elected to serve additional 2-year terms. The same process will be used in the event that a **TSC Member** resigns prior to the end of their term.
```

# OpenQASM Governance model

## Purpose

The purpose of the OpenQASM governance model is to formalize the process for proposing and deciding upon changes to OpenQASM. It provides explicit mechanisms for a broader community to contribute. It will memorialize decisions in a public record. This model should make decision-making routine to get issues unstuck.

> OpenQASM is a Qiskit project, but this governance structure is scoped to just OpenQASM itself and not Qiskit as a whole. The scope of OpenQASM includes the language specification as well as reference implementations of parsers or other low-level infrastructure to facilitate use in other projects.

Author: To clarify what it means to be a "Qiskit project":
```suggestion
This governance structure is scoped to just OpenQASM itself and not Qiskit as a whole. The scope of OpenQASM includes the language specification as well as reference implementations of parsers or other low-level infrastructure to facilitate use in other projects. OpenQASM inherits the contribution guidelines and community standards that are common to all projects within the Qiskit GitHub organization.
```

**In brief**: we propose 3 community roles: **Member**, **Contributor**, and **TSC Member**. We also propose 2 structures: a **Technical Steering Committee** (TSC) and **Working Group** (WG). The TSC creates WGs and appoints WG chairs. Working Group membership shall be composed of Members with a Contributor as chair.

## Roles

**Member** - individuals that participate in the OpenQASM project and community. These people must adhere to the Qiskit Code of Conduct but have no specific further responsibilities.

> **Contributor** - a **Member** who is an active contributor to the OpenQASM community. This person has a *sustained history* of authoring, commenting on, and reviewing issues and PRs. They post questions or responses to the community Slack (#open-qasm in the Qiskit workspace). Contributors are selected by nomination of at least 2 current TSC Members. *Contributors may participate in WGs and attend TSC meetings.*

Reviewer 2: Clarify that contributors are non-voting.

Author: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

**TSC Member** - a **Contributor** who also regularly attends and has voting rights at Technical Steering Committee meetings.

## Structures

We will form a **Technical Steering Committee**. The TSC shall be composed of 5 Contributors. The initial TSC will include:
* Ali Javadi (IBM Quantum)
* Lev S Bishop (IBM Quantum)
* Blake Johnson (IBM Quantum)
* Steven Heidel (AWS)
* TBD (another non-IBM member, to be selected by the TSC)

> TSC meetings shall be open to all **Contributors**. The TSC will make decisions by simple majority vote of its members. TSC meetings shall require a quorum of 60% of voting members present to make a decision by vote (ie, 3 members for the initial TSC). If a TSC meeting does not have quorum, it may still discuss and record minutes. The TSC may make decisions outside of meetings (e.g. on a pull request) so long as voting/approval in that forum meets the quorum requirement. Upon request from any TSC member, any decision must be deferred until the next TSC meeting, to enable live discussion.

Reviewer 2: Two suggestions for this paragraph:
1. I would like to insert some language that the TSC aims to build consensus among contributors prior to voting in most cases. Voting to resolve disagreements is ideally expected to happen less often than voting to accept the consensus. There is no way to write this into the structure but it's a goal we should strive towards.
2. Allow the TSC to invite guests to the TSC meetings, even if they are not contributors yet.

Author: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

> TSC meetings shall be open to all **Contributors**. The TSC will make decisions by simple majority vote of its members. TSC meetings shall require a quorum of 60% of voting members present to make a decision by vote (ie, 3 members for the initial TSC). If a TSC meeting does not have quorum, it may still discuss and record minutes. The TSC may make decisions outside of meetings (e.g. on a pull request) so long as voting/approval in that forum meets the quorum requirement. Upon request from any TSC member, any decision must be deferred until the next TSC meeting, to enable live discussion.

Author: It effectively postpones the vote until the meeting. TSC members might then choose to change their votes.

> The TSC responsibility is to balance a desire to allow a voice to the broad OpenQASM community; with lightweight processes to enable moving fast and maintaining a consistent philosophy for the language. Thus, for example, this governance document does give the freedom to the TSC to make decisions without calling a meeting and explicitly involving contributors, but the idea is that the TSC should make use of this power judiciously. The TSC has authority to make all decisions regarding OpenQASM (including amendments to this governance document), but should apply more process and seek more community input on constitutional changes than for simple fixes to the spec.

Reviewer 3: ```suggestion
The TSC responsibility is to balance a desire to allow a voice to the broad OpenQASM community with lightweight processes to enable moving fast and maintaining a consistent philosophy for the language. Thus, for example, this governance document does give the freedom to the TSC to make decisions without calling a meeting and explicitly involving contributors, but the idea is that the TSC should make use of this power judiciously. The TSC has authority to make all decisions regarding OpenQASM (including amendments to this governance document), but should apply more process and seek more community input on constitutional changes than for simple fixes to the spec.
```

> The TSC may have a **Secretary** (non-voting **Contributor**) who organizes the meeting (e.g. prepares and distributes the agenda and collects a set of issues to be considered). The TSC **Secretary** will be appointed by vote of the TSC.

Author: The secretary cannot break ties, but they can participate in the discussion.

> The TSC may have a **Secretary** (non-voting **Contributor**) who organizes the meeting (e.g. prepares and distributes the agenda and collects a set of issues to be considered). The TSC **Secretary** will be appointed by vote of the TSC.

Reviewer 2: ```suggestion
The TSC may have a **Secretary** who organizes the meeting (e.g. prepares and distributes the agenda and collects a set of issues to be considered). The TSC **Secretary** will be a **Contributor** appointed by vote of the TSC.
```

"non-voting Contributor" is confusing since all contributors are non-voting.

Author: Addressed in 70baf4cbd88f9705d78206d061f3e16b6614b8a7

* Zach Schoenfeld (IBM Quantum) is nominated to serve as the first TSC **Secretary**.

> The TSC will, from time to time, elect to create **Working Groups** to study issues. The TSC will appoint a **Contributor** as chair for each WG. These WGs should come back to the TSC with a proposal in the form of a language RFC or pull request to the OpenQASM specification. A WG will be automatically disbanded upon acceptance or (final) rejection of an RFC. A potential set of initial Working Groups might include:

Reviewer 3: Is there a centralized location where WG's will be listed? I think it would be beneficial to the community to at least know what issues are currently being explored...

Reviewer 2: Would be good to add this to the README and also the live spec.

Rust does a really good job at this: https://www.rust-lang.org/governance

Author: Wow, that is nice. Having something as fancy would clearly take infrastructure we don't have yet, but listing WGs in the README would be a good start.

* Pulse grammar
* Types and casting (including implicit casts)
* Generics, circuit families, and subroutines
* Defcal for non-unitaries

## Open Questions

> * How will new **TSC Members** be selected when current **TSC Members** step down?

Reviewer 4: 👍

    - Suggestion: nomination power to reside with TSC, voting open to all **Contributors**

> * Do we need term limits? The various source models we considered typically use a 2 year term.

Reviewer 4: Maybe a term of office, and not a term limit? Eg., A TSC member is appointed for two-year term and possibly reelected/affirmed at the end of this term to continue another two-year term. However, there is no limit to the number of terms a TSC member may serve?

Author: Yes, a term of office would be better.

Reviewer 3: I think this suggestion makes sense...

Author: If we do terms, I suggest staggering the initial terms so that the election process isn't choosing an entire TSC at once.
